File: src/D2BotNG/Rendering/Dc6Decoder.cs
using System.Runtime.InteropServices;

namespace D2BotNG.Rendering;

/// <summary>
/// DC6 file format header structure
/// </summary>
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct Dc6Header
{
    public int Version;
    public int SubVersion;
    public int Zeros;
    public int Termination;
    public int Directions;
    public int FramesPerDirection;
}

/// <summary>
/// DC6 frame header structure
/// </summary>
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct Dc6FrameHeader
{
    public int Flip;
    public int Width;
    public int Height;
    public int OffsetX;
    public int OffsetY;
    public int Zeros;
    public int NextBlock;
    public int Length;
}

/// <summary>
/// Decoded DC6 frame data
/// </summary>
public class Dc6Frame
{
    public int Width { get; init; }
    public int Height { get; init; }
    public byte[,] Pixels { get; init; } = null!;
}

/// <summary>
/// Decodes Diablo 2 DC6 sprite files
/// </summary>
public static class Dc6Decoder
{
    private const int HeaderSize = 24; // sizeof(Dc6Header)
    private const int FrameHeaderSize = 32; // sizeof(Dc6FrameHeader)

    /// <summary>
    /// Decodes the first frame from a DC6 file
    /// </summary>
    public static Dc6Frame DecodeFirstFrame(byte[] dc6Data)
    {
        if (dc6Data.Length < HeaderSize)
            throw new ArgumentException("DC6 data too small for header");

        // Read main header
        var header = ReadStruct<Dc6Header>(dc6Data, 0);

        if (header.Directions < 1 || header.FramesPerDirection < 1)
            throw new ArgumentException("Invalid DC6 header: no frames");

        // Read first frame pointer (located after main header)
        int framePointer = BitConverter.ToInt32(dc6Data, HeaderSize);

        // Read frame header
        var frameHeader = ReadStruct<Dc6FrameHeader>(dc6Data, framePointer);

        // Decode the frame pixels
        var pixels = DecodeFramePixels(dc6Data, framePointer + FrameHeaderSize, frameHeader);

        return new Dc6Frame
        {
            Width = frameHeader.Width,
            Height = frameHeader.Height,
            Pixels = pixels
        };
    }

    /// <summary>
    /// Decodes all frames from a DC6 file (used for fonts)
    /// </summary>
    // ReSharper disable once UnusedMember.Global — rendering utility for future font support
    public static Dc6Frame[] DecodeAllFrames(byte[] dc6Data)
    {
        if (dc6Data.Length < HeaderSize)
            throw new ArgumentException("DC6 data too small for header");

        var header = ReadStruct<Dc6Header>(dc6Data, 0);
        int totalFrames = header.Directions * header.FramesPerDirection;

        var frames = new Dc6Frame[totalFrames];

        for (int i = 0; i < totalFrames; i++)
        {
            int framePointer = BitConverter.ToInt32(dc6Data, HeaderSize + i * 4);
            var frameHeader = ReadStruct<Dc6FrameHeader>(dc6Data, framePointer);
            var pixels = DecodeFramePixels(dc6Data, framePointer + FrameHeaderSize, frameHeader);

            frames[i] = new Dc6Frame
            {
                Width = frameHeader.Width,
                Height = frameHeader.Height,
                Pixels = pixels
            };
        }

        return frames;
    }

    /// <summary>
    /// Decodes the RLE-compressed pixel data from a DC6 frame
    /// </summary>
    private static byte[,] DecodeFramePixels(byte[] data, int dataOffset, Dc6FrameHeader header)
    {
        var pixels = new byte[header.Width, header.Height];

        if (header.Width <= 0 || header.Height <= 0)
            return pixels;

        int x = 0;
        int y = header.Height - 1; // DC6 is stored bottom-to-top

        for (int bytesRead = 0; bytesRead < header.Length && dataOffset < data.Length;)
        {
            byte b = data[dataOffset++];
            bytesRead++;

            if (b == 0x80) // Row terminator
            {
                x = 0;
                y--;
                if (y < 0) break;
            }
            else if ((b & 0x80) == 0x80) // Skip pixels (transparent)
            {
                x += b & 0x7F;
            }
            else // Literal run
            {
                int count = b;
                for (int i = 0; i < count && dataOffset < data.Length; i++)
                {
                    if (x < header.Width && y >= 0)
                    {
                        pixels[x, y] = data[dataOffset];
                    }
                    dataOffset++;
                    bytesRead++;
                    x++;
                }
            }
        }

        return pixels;
    }

    private static T ReadStruct<T>(byte[] data, int offset) where T : struct
    {
        int size = Marshal.SizeOf<T>();
        if (offset + size > data.Length)
            throw new ArgumentException($"Not enough data to read {typeof(T).Name}");

        nint ptr = Marshal.AllocHGlobal(size);
        try
        {
            Marshal.Copy(data, offset, ptr, size);
            return Marshal.PtrToStructure<T>(ptr);
        }
        finally
        {
            Marshal.FreeHGlobal(ptr);
        }
    }
}

File: src/D2BotNG.UI/src/lib/rendering/dc6Decoder.ts
/**
 * DC6 file format decoder for Diablo 2 sprites
 */

export interface Dc6Frame {
  width: number;
  height: number;
  offsetX: number;
  offsetY: number;
  pixels: Uint8Array; // Palette indices, row-major [y * width + x]
}

interface Dc6Header {
  version: number;
  subVersion: number;
  directions: number;
  framesPerDirection: number;
}

interface Dc6FrameHeader {
  flip: number;
  width: number;
  height: number;
  offsetX: number;
  offsetY: number;
  nextBlock: number;
  length: number;
}

const HEADER_SIZE = 24;
const FRAME_HEADER_SIZE = 32;

function readInt32(data: DataView, offset: number): number {
  return data.getInt32(offset, true); // little-endian
}

function readHeader(data: DataView): Dc6Header {
  return {
    version: readInt32(data, 0),
    subVersion: readInt32(data, 4),
    // zeros at 8
    // termination at 12
    directions: readInt32(data, 16),
    framesPerDirection: readInt32(data, 20),
  };
}

function readFrameHeader(data: DataView, offset: number): Dc6FrameHeader {
  return {
    flip: readInt32(data, offset),
    width: readInt32(data, offset + 4),
    height: readInt32(data, offset + 8),
    offsetX: readInt32(data, offset + 12),
    offsetY: readInt32(data, offset + 16),
    // zeros at offset + 20
    nextBlock: readInt32(data, offset + 24),
    length: readInt32(data, offset + 28),
  };
}

function decodeFramePixels(
  data: Uint8Array,
  dataOffset: number,
  header: Dc6FrameHeader,
): Uint8Array {
  const pixels = new Uint8Array(header.width * header.height);

  if (header.width <= 0 || header.height <= 0) {
    return pixels;
  }

  let x = 0;
  let y = header.height - 1; // DC6 is stored bottom-to-top
  let offset = dataOffset;
  let bytesRead = 0;

  while (bytesRead < header.length && offset < data.length) {
    const b = data[offset++];
    bytesRead++;

    if (b === 0x80) {
      // Row terminator
      x = 0;
      y--;
      if (y < 0) break;
    } else if ((b & 0x80) === 0x80) {
      // Skip pixels (transparent)
      x += b & 0x7f;
    } else {
      // Literal run
      const count = b;
      for (let i = 0; i < count && offset < data.length; i++) {
        if (x < header.width && y >= 0) {
          pixels[y * header.width + x] = data[offset];
        }
        offset++;
        bytesRead++;
        x++;
      }
    }
  }

  return pixels;
}

/**
 * Decodes the first frame from DC6 data
 */
export function decodeFirstFrame(buffer: ArrayBuffer): Dc6Frame {
  const data = new Uint8Array(buffer);
  const view = new DataView(buffer);

  if (data.length < HEADER_SIZE) {
    throw new Error("DC6 data too small for header");
  }

  const header = readHeader(view);

  if (header.directions < 1 || header.framesPerDirection < 1) {
    throw new Error("Invalid DC6 header: no frames");
  }

  // Read first frame pointer (located after main header)
  const framePointer = readInt32(view, HEADER_SIZE);

  // Read frame header
  const frameHeader = readFrameHeader(view, framePointer);

  // Decode the frame pixels
  const pixels = decodeFramePixels(
    data,
    framePointer + FRAME_HEADER_SIZE,
    frameHeader,
  );

  return {
    width: frameHeader.width,
    height: frameHeader.height,
    offsetX: frameHeader.offsetX,
    offsetY: frameHeader.offsetY,
    pixels,
  };
}

/**
 * Decodes all frames from DC6 data
 */
export function decodeAllFrames(buffer: ArrayBuffer): Dc6Frame[] {
  const data = new Uint8Array(buffer);
  const view = new DataView(buffer);

  if (data.length < HEADER_SIZE) {
    throw new Error("DC6 data too small for header");
  }

  const header = readHeader(view);
  const totalFrames = header.directions * header.framesPerDirection;
  const frames: Dc6Frame[] = [];

  for (let i = 0; i < totalFrames; i++) {
    const framePointer = readInt32(view, HEADER_SIZE + i * 4);
    const frameHeader = readFrameHeader(view, framePointer);
    const pixels = decodeFramePixels(
      data,
      framePointer + FRAME_HEADER_SIZE,
      frameHeader,
    );

    frames.push({
      width: frameHeader.width,
      height: frameHeader.height,
      offsetX: frameHeader.offsetX,
      offsetY: frameHeader.offsetY,
      pixels,
    });
  }

  return frames;
}
